### PR TITLE
Fix bug with SQLGetData

### PR DESCRIPTION
**Pull request by forestmvey**
### Summary

SQLGetData was returning SQL_NO_TOTAL which cause PowerBI ODBC and custom connection failures when data returned is empty or finished reading data.

### Description

To align with [SQLGetData function](https://learn.microsoft.com/en-us/sql/odbc/reference/syntax/sqlgetdata-function?view=sql-server-ver16#retrieving-data-with-sqlgetdata), SQL_NO_TOTAL should be set only when response size is unknown. The login timeout will be handled in follow up work and for now the value is ignored.

### Related Issue

[Issue 19](https://github.com/awslabs/amazon-timestream-odbc-driver/issues/19)

### Validations
- [x] IT tests pass
- [x] UT tests pass
- [x] Direct query with PowerBI custom connector
- [x] Import with PowerBI and ODBC
- [x] Tested C# application using OdbcConnection on MSDN
```
using System.Diagnostics;
using System;
using System.Data.Odbc;

class Program { 
    
    static void Main(string[] args) { 
        
        Debug.Listeners.Add(new ConsoleTraceListener());
        string queryString = "SELECT * FROM data_queries_test_db.TestComplexTypes";

        using (OdbcConnection connection = new OdbcConnection("DSN=Timestream DSN;"))
        {
            OdbcCommand command = new OdbcCommand(queryString, connection);
            connection.Open();
            // Execute the DataReader and access the data.
            OdbcDataReader reader = command.ExecuteReader();
            while (reader.Read())
            {
                Debug.WriteLine("{0}", reader[0]);
            }
            reader.Close();
        }
        Console.ReadLine();
    } 
} 
```
**Compile code**:
```
csc  /define:DEBUG .\odbcTest.cs
```

- [x] Tested ADODB powershell script query
```
# Define the connection string using the MSDASQL provider for ODBC
# Replace 'YourDSN' with your Data Source Name configured in ODBC Data Source Administrator

$connectionString = "DSN=Timestream DSN;logLevel=4;"
$query = "SELECT * FROM data_queries_test_db.TestComplexTypes"

# ------------------- OLE DB -------------------
$connection = New-Object -ComObject ADODB.Connection
$connection.Open($connectionString)

Write-Host -NoNewline 'Press any key to continue . . .';
$null = $Host.UI.RawUI.ReadKey('NoEcho,IncludeKeyDown');

# Create the command object
$command = New-Object -ComObject ADODB.Command
$command.ActiveConnection = $connection
$command.CommandText = $query

# Execute the query
$recordset = $command.Execute()

# Loop through results and display them
if ($recordset) {
	while (!$recordset.EOF) {
		$rowData = ""
		for ($i = 0; $i -lt $recordset.Fields.Count; $i++) {
			$field = $recordset.Fields.Item($i)
			$fieldValue = $field.Value
			if ($fieldValue -is [System.DBNull]) {
				$fieldValue = "-"
			}
			$rowData += "$($field.Name): $($fieldValue)"
			if ($i -lt $recordset.Fields.Count - 1) {
				$rowData += " | "
			}
		}
		Write-Output $rowData
		Write-Output ""
    	$recordset.MoveNext()
	}
	$recordset.Close()
}

$connection.Close()

```
- [x] Test with Excel using OLE DB
Screenshot 2024-09-11 at 2 21 08 PM

### CODE REVIEW

File: src/tests/integration-test/src/api_robustness_test.cpp [hunk 1]
```diff
@@ -315,6 +315,25 @@ BOOST_AUTO_TEST_CASE(TestSQLFetchPastEnd) {
   SQLCloseCursor(stmt);
 }
 
+BOOST_AUTO_TEST_CASE(TestSQLFetchEmpty) {
```
Comment (jduo) [inline src/tests/integration-test/src/api_robustness_test.cpp:318]: This test case seems unrelated to the fix which is about SQLGetData

Comment (forestmvey) [reply src/tests/integration-test/src/api_robustness_test.cpp:318]: Previously this would set SQL_NO_TOTAL, which is not valid.

Comment (jduo) [reply src/tests/integration-test/src/api_robustness_test.cpp:318]: This test is getting the return value for SQLFetch though, not SQLGetData.

Comment (forestmvey) [reply src/tests/integration-test/src/api_robustness_test.cpp:318]: Wups, I thought I saw the test failing. I will isolate SQLGetData in another test.

Comment (RoyZhang2022) [reply src/tests/integration-test/src/api_robustness_test.cpp:318]: I did not see your SQLGetData test.

Comment (forestmvey) [reply src/tests/integration-test/src/api_robustness_test.cpp]: Added the test for `SQLGetData` with an empty response.

File: src/odbc/src/app/application_data_buffer.cpp [hunk 2]
```diff
@@ -282,7 +282,7 @@ ConversionResult::Type ApplicationDataBuffer::PutStrToStrBuffer(
   SqlUlen inCharIndex = currentCellOffset / inCharSize;
 
   if (inCharIndex >= value.length()) {
-    if (resLenPtr) {
+    if (resLenPtr && value.length() != 0) {
       *resLenPtr = SQL_NO_TOTAL;
```
Comment (jduo) [inline src/odbc/src/app/application_data_buffer.cpp:286]: I still feel like this is wrong...?
SQL_NO_TOTAL should only ever be returned if there is more data in the cell (ie it got truncated) but this looks like it occurs if we are calling SQLGetData past the end of the cell.

Comment (jduo) [reply src/odbc/src/app/application_data_buffer.cpp:286]: Really, Line 335 is the scenario where SQL_NO_TOTAL should be returned, but the remaining length is already calculated so it's not necessary.

Comment (RoyZhang2022) [reply src/odbc/src/app/application_data_buffer.cpp:286]: Yes, this looks like more data in the cell. It is not the return size unknown.

File: src/odbc/src/app/application_data_buffer.cpp [hunk 3]
```diff
@@ -278,15 +279,14 @@ ConversionResult::Type ApplicationDataBuffer::PutStrToStrBuffer(
...
   SqlUlen currentCellOffset = cellOffset >= 0 ? cellOffset : 0;
-  // Since cellOffset is in bytes, an index needs to be calculated.
-  SqlUlen inCharIndex = currentCellOffset / inCharSize;
 
-  if (inCharIndex >= value.length()) {
-    if (resLenPtr) {
-      *resLenPtr = SQL_NO_TOTAL;
-    }
+  // If all data has already been read, return AI_NO_DATA
+  if ((bytesRequired - ((currentCellOffset / inCharSize) * outCharSize)) <= 0) {
```
Comment (jduo) [inline src/odbc/src/app/application_data_buffer.cpp:284]: This check should be refactored because both bytesRequired and currentCellOffset are unsigned integers so the result of subtraction can't be < 0.

File: src/odbc/src/app/application_data_buffer.cpp [hunk 4]
```diff
@@ -310,6 +310,20 @@ ConversionResult::Type ApplicationDataBuffer::PutStrToStrBuffer(
     assert(false);
   }
 
+  if (isTruncated) {
+      if (resLenPtr) {
+        *resLenPtr = SQL_NO_TOTAL;
```
Comment (jduo) [inline src/odbc/src/app/application_data_buffer.cpp:315]: While this is correct, it's actually more informative to the app to report the exact remaining bytes (which I believe we calculated earlier in the function). That lets the app create a buffer of the appropriate size instead of creating an unknown-sized buffer.

Comment (forestmvey) [reply src/odbc/src/app/application_data_buffer.cpp]: I have updated to not use SQL_NO_TOTAL!

File: src/odbc/src/app/application_data_buffer.cpp [hunk 5]
```diff
@@ -278,15 +279,14 @@ ConversionResult::Type ApplicationDataBuffer::PutStrToStrBuffer(
...
-    if (resLenPtr) {
-      *resLenPtr = SQL_NO_TOTAL;
-    }
+  // If all data has already been read, return AI_NO_DATA
+  if (((currentCellOffset / inCharSize) * outCharSize) >= bytesRequired) {
     return ConversionResult::Type::AI_NO_DATA;
   }
+  
+  // Since cellOffset is in bytes, an index needs to be calculated.
+  SqlUlen inCharIndex = currentCellOffset / inCharSize;
```
Comment (jduo) [inline src/odbc/src/app/application_data_buffer.cpp:289]: This gets calculated twice. Here and at line 284.
Also please expand the names (like specify that the index is in code units or encoding, if it's the source or target, etc)

File: src/odbc/src/app/application_data_buffer.cpp [hunk 6]
```diff
@@ -326,14 +326,13 @@ ConversionResult::Type ApplicationDataBuffer::PutStrToStrBuffer(
...
-    size_t numCharsWritten = bytesWritten / outCharSize;
-    SetCellOffset(cellOffset + numCharsWritten * inCharSize);
-  }
-
   if (isTruncated) {
+    SetCellOffset(currentCellOffset + (buflen / outCharSize));
     return ConversionResult::Type::AI_VARLEN_DATA_TRUNCATED;
   } else {
+    if (cellOffset >= 0) {
+      SetCellOffset(totalBytesWritten);
```
Comment (jduo) [inline src/odbc/src/app/application_data_buffer.cpp:334]: Just want to make sure I understand this correctly. This sets the cell offset in bytes right? And totalBytesWritten does _not_ include bytes for the NUL terminator.

Comment (forestmvey) [reply src/odbc/src/app/application_data_buffer.cpp:334]: Correct, it does not include the bytes for the NUL terminator.

File: src/odbc/src/app/application_data_buffer.cpp [hunk 7]
```diff
@@ -256,9 +256,10 @@ ConversionResult::Type ApplicationDataBuffer::PutStrToStrBuffer(
   LOG_DEBUG_MSG("inCharSize is " << inCharSize << ", outCharSize is "
                                  << outCharSize << ", buflen is " << buflen);
 
+  size_t valueLength = value.length();
   size_t bytesRequired = 0;
   if (ANSI_STRING_ONLY) {
-    bytesRequired = value.length() * outCharSize;
+    bytesRequired = valueLength * outCharSize;
   } else {
     thread_local std::wstring_convert<std::codecvt_utf8<wchar_t >, wchar_t>
```
[the remark this answers is not in the document]
Comment (forestmvey) [reply src/odbc/src/app/application_data_buffer.cpp:264]: I can create a follow up ticket to address the conversion issue.

File: src/odbc/src/app/application_data_buffer.cpp (hunk 5 above)
Comment (forestmvey) [reply src/odbc/src/app/application_data_buffer.cpp]: Updated. to not re-calculate the index, in fact it's a meaningless call anyways. inCharSize is always 1, so all we care about is the currentCellOffset.

**Review by jduo — APPROVED** [review]
